mlflow/tensorflow/__init__.py

```diff
@@ -888,6 +996,99 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
+                input_training_data = input_training_data[:][0]
+
+            if isinstance(input_training_data, np.ndarray):
+                input_example_slice = input_training_data[:INPUT_EXAMPLE_SAMPLE_ROWS]
+            elif isinstance(input_training_data, tensorflow.data.Dataset):
+                steps = 1
+                if history.params is not None and "steps" in history.params:
+                    steps = history.params["steps"]
+
+                def _extract_n_steps(input_example_n_steps):
+                    # Example:
+                    # when the tensorflow.data.Dataset is batched
+                    # (Total number of steps (batches of samples)),
+                    # extract first ``INPUT_EXAMPLE_SAMPLE_ROWS`` samples from the first batch
+                    if steps > 1:
+                        return np.array(
+                            [
+                                v[0][:INPUT_EXAMPLE_SAMPLE_ROWS]
+                                for v in input_example_n_steps.take(1).as_numpy_iterator()
+                            ]
+                        )[0]
+                    return np.array(
+                        [
+                            v[0]
+                            for v in input_example_n_steps.take(
+                                INPUT_EXAMPLE_SAMPLE_ROWS
+                            ).as_numpy_iterator()
+                        ]
+                    )
+
+                input_example_slice = _extract_n_steps(input_training_data)
+            elif isinstance(input_training_data, dict):
+                input_example_slice = {
+                    k: np.take(v, range(0, INPUT_EXAMPLE_SAMPLE_ROWS))
+                    for k, v in input_training_data.items()
+                }
+            else:
+                warnings.warn(
+                    "Tensorflow keras autologging only "
+                    "supports input types of: numpy.ndarray, "
+                    "dict(<key> -> numpy.ndarray), tensorflow.data.Dataset, "
+                    "or tensorflow.keras.utils.Sequence"
+                )
+
+            return input_example_slice
+
+        def _infer_model_signature(input_data_slice):
+            original_stop_training = history.model.stop_training
+            model_output = history.model.predict(input_data_slice)
+
+            if Version(tensorflow.__version__) <= Version("2.1.4") and original_stop_training:
+                # For these versions, `stop_training` flag on Model is set to False
+                # This flag is used by the callback
+                # (inside ``_log_early_stop_callback_metrics``)
+                # for logging of early stop metrics. In order for
+                # that to work, need to force that flag to be True again since doing
+                # predict on that model sets `stop_training` to false for
+                # those TF versions
+                history.model.stop_training = True
```

Comment:
I think it would be simpler and version-agnostic to do the following:

```suggestion
            original_stop_training = history.model.stop_training
            model_output = history.model.predict(input_data_slice)
            history.model.stop_training = original_stop_training
```

Reply:
this change is in master now.